packages/protocol/engine/tangle/booker/attachments.go

```diff
@@ -43,57 +118,66 @@ func (a *attachments) Evict(epochIndex epoch.Index) {
 		a.evictionMap.Delete(epochIndex)
 
 		txIDs.ForEach(func(txID utxo.TransactionID) {
+			a.mutex.Lock(txID)
+			defer a.mutex.Unlock(txID)
+
 			if attachmentsOfTX := a.storage(txID, false); attachmentsOfTX != nil && attachmentsOfTX.Delete(epochIndex) && attachmentsOfTX.Size() == 0 {
 				a.attachments.Delete(txID)
 			}
 		})
 	}
 }
 
-func (a *attachments) storeAttachment(txID utxo.TransactionID, block *Block) {
-	attachmentsOfEpoch, _ := a.storage(txID, true).RetrieveOrCreate(block.ID().EpochIndex, func() *LockableSlice[*Block] {
-		return NewLockableSlice[*Block]()
+func (a *attachments) storeAttachment(txID utxo.TransactionID, block *Block) (created bool) {
+	attachmentsOfEpoch, _ := a.storage(txID, true).RetrieveOrCreate(block.ID().EpochIndex, func() *memstorage.Storage[models.BlockID, *Block] {
+		return memstorage.New[models.BlockID, *Block]()
 	})
-	attachmentsOfEpoch.Append(block)
+
+	return lo.Return2(attachmentsOfEpoch.RetrieveOrCreate(block.ID(), func() *Block {
+		return block
+	}))
 }
 
-func (a *attachments) getEarliestAttachment(txID utxo.TransactionID) (attachment *Block) {
+func (a *attachments) getEarliestAttachment(txID utxo.TransactionID, returnOrphaned ...bool) (attachment *Block) {
 	var lowestTime time.Time
 	if txStorage := a.storage(txID, false); txStorage != nil {
-		txStorage.ForEach(func(_ epoch.Index, blocks *LockableSlice[*Block]) bool {
-			for _, block := range blocks.Slice() {
-				if lowestTime.After(block.IssuingTime()) || lowestTime.IsZero() {
+		txStorage.ForEach(func(_ epoch.Index, blocks *memstorage.Storage[models.BlockID, *Block]) bool {
+			blocks.ForEach(func(_ models.BlockID, block *Block) bool {
+				if (lowestTime.After(block.IssuingTime()) || lowestTime.IsZero()) && (!block.IsOrphaned() || (len(returnOrphaned) > 0 && returnOrphaned[0])) {
 					lowestTime = block.IssuingTime()
 					attachment = block
 				}
-			}
+
+				return true
+			})
 			return true
 		})
 	}
 
 	return
 }
 
-func (a *attachments) getLatestAttachment(txID utxo.TransactionID) (attachment *Block) {
+func (a *attachments) getLatestAttachment(txID utxo.TransactionID, returnOrphaned ...bool) (attachment *Block) {
```

Comment:
Check timestamp monotonicity for the attachment to be valid? The check should either be here or in `referenceProvider.latestValidAttachment`